src/Features/LanguageServer/ProtocolUnitTests/DocumentChanges/DocumentChangesTests.cs

```diff
@@ -0,0 +1,367 @@
+
+                await DidChange(queue, workspace.CurrentSolution, CreateDidChangeTextDocumentParams(locationTyped.Uri, (4, 8, "// hi there")));
+
+                var documentTextFromWorkspace = (await workspace.CurrentSolution.GetDocuments(locationTyped.Uri).Single().GetTextAsync()).ToString();
+
+                Assert.NotNull(documentTextFromWorkspace);
+                Assert.Equal(documentText, documentTextFromWorkspace);
+
+                // Just to ensure this test breaks if didChange stops working for some reason
+                Assert.NotEqual(expected, documentTextFromWorkspace);
+            }
+        }
+
+        [Fact]
+        public async Task DidChange_MultipleChanges()
+        {
+            var source =
+@"class A
+{
+    void M()
+    {
+        {|type:|}
+    }
+}";
+            var expected =
+  @"class A
+{
+    void M()
+    {
+        // hi there
+        // this builds on that
+    }
+}";
+
+            var (workspace, locationTyped, documentText) = await GetWorkspaceAndLocationAsync(source);
+
+            using (workspace)
+            {
+                var queue = CreateRequestQueue(workspace.CurrentSolution);
+
+                await DidOpen(queue, workspace.CurrentSolution, CreateDidOpenTextDocumentParams(locationTyped, documentText));
+
+                await DidChange(queue, workspace.CurrentSolution, CreateDidChangeTextDocumentParams(locationTyped.Uri, (4, 8, "// hi there"), (5, 0, "        // this builds on that\r\n")));
+
+                var document = queue.GetTestAccessor().GetTrackedDocuments().FirstOrDefault();
+
+                Assert.NotNull(document);
+                Assert.Equal(expected, document.GetTextSynchronously(default).ToString());
+            }
+        }
+
+        [Fact]
+        public async Task DidChange_MultipleRequests()
+        {
+            var source =
+@"class A
+{
+    void M()
+    {
+        {|type:|}
+    }
+}";
+            var expected =
+  @"class A
+{
+    void M()
+    {
+        // hi there
+        // this builds on that
+    }
+}";
+
+            var (workspace, locationTyped, documentText) = await GetWorkspaceAndLocationAsync(source);
+
+            using (workspace)
+            {
+                var queue = CreateRequestQueue(workspace.CurrentSolution);
+
+                await DidOpen(queue, workspace.CurrentSolution, CreateDidOpenTextDocumentParams(locationTyped, documentText));
+
+                await DidChange(queue, workspace.CurrentSolution, CreateDidChangeTextDocumentParams(locationTyped.Uri, (4, 8, "// hi there")));
+
+                await DidChange(queue, workspace.CurrentSolution, CreateDidChangeTextDocumentParams(locationTyped.Uri, (5, 0, "        // this builds on that\r\n")));
+
+                var document = queue.GetTestAccessor().GetTrackedDocuments().FirstOrDefault();
+
+                Assert.NotNull(document);
+                Assert.Equal(expected, document.GetTextSynchronously(default).ToString());
+            }
+        }
```

Comment:
Can we add a full roundtrip test that asserts the state of the document and document tracker at all steps?

i.e.
1.  open (asserts document is in the tracker with correct text)
2.  change (asserts document is in tracker with updated text)
3.  close (asserts document is no longer in the tracker)

Reply:
Added